enhancements/cloud-controller-manager/openstack-cloud-controller-manager.md

```diff
@@ -0,0 +1,278 @@
+
+- On the initial stage of upgrading `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+
+- `cluster-cloud-controller-manager-operator` creates all required resources for CCM (Namespace, RBAC, Service Account).
+
+- `cluster-cloud-controller-manager-operator` creates a static pod manifest for CCM.
+
+- `kubelet` is restarted to start CCM in a static pod.
+
+- `kube-apiserver` and `kube-controller-manager` are restarted without the `--cloud-provider` option.
+
+- `kubelet` is restarted with `--cloud-provider external` option.
+
+Downgrade:
+
+- `kubelet` is restarted with `--cloud-provider openstack` option.
+
+- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider openstack` option.
+
+- `cluster-cloud-controller-manager-operator` destroys OpenStack CCM manifests and all related resources.
+
+- `kubelet` is restarted again.
+
+- `cluster-cloud-controller-manager-operator` stops working.
+
+### Version Skew Strategy
+
+See the upgrade/downgrade strategy.
+
+## Alternatives
+
+Only one alternative is to keep the deprecated in-tree cloud providers in OpenShift and support them with our own resources.
+
+## Infrastructure Needed
+
+Additional OpenStack cloud may be required to test how CCM works with self-signed certificates. Our current CI doesn't allow this.
+
+## Open questions
+
+1. Should we reuse the existing cloud provider config or generate a new one?
+CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.
+
+2. How to migrate PVs created by the in-tree cloud provider?
+[CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.
+
+## FAQ
+
+Q: Can we disaster-recover a cluster without CCM running?
+A: We can't. Basically, without CCM running, nodes can only join the cluster, but they will be unschedulable. Which means it's impossible to start any workloads on the nodes. [Source](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager)
+
+Q: What is non-functional in a cluster during bootstrapping until CCM is up?
+A: If CCM is not available, new nodes in the cluster will be left unschedulable.
+
+Q: Who talks to CCM?
+A: From architectural point of view, only Kube API server is able to communicate with CCM. [Source](https://kubernetes.io/docs/concepts/architecture/cloud-controller/#design). All requests to CCM are sent by `kubelet`. KCM doesn't interract with CCM.
+
+Q: Does CCM provide an API? How is it hosted? HTTPS endpoint? Through load balancer?
+A: CCM provides a gRPC interface on port [10258](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/ports.go#L22). Like KCM, CCM supports high available setup using leader election (which is on by default). [Source](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#requirements).
```

Comment:
this means that clients of ccm (kubelet? kube-apiserver?) have to find the leader? I.e. you need a load balancer? Or is it a kubernetes service?

Reply:
In the former case you have to integrate with the cloud API LB, and also integrate with MCO and it gcp/azure-routes.sh mechanism. Those might even be duplicated because API server and ccm life-cycles don't match.

In the latter case, you depend on SDN. What would that mean? No scheduling, no SDN. No SDN, no ccm, no scheduling.

Reply:
Or you tell kube-apiserver about all the IPs similarly to etcd-operator today. Grpc can handle multiple endpoints and find one that works.